prow/test/integration/fakegitserver/fakegitserver.go

```diff
@@ -0,0 +1,325 @@
+		return nil, err
+	}
+
+	if err := setGitConfigOptions(repo); err != nil {
+		logrus.Errorf("config setup failed")
+		return nil, err
+	}
+
+	if err := runSetupScript(dir, repoSetup.Script); err != nil {
+		logrus.Errorf("running the repo setup script failed")
+		return nil, err
+	}
+
+	if err := convertToBareRepo(repo, dir); err != nil {
+		logrus.Errorf("conversion to bare repo failed")
+		return nil, err
+	}
+
+	repo, err = git.PlainOpen(dir)
+	if err != nil {
+		logrus.Errorf("could not reopen repo")
+		return nil, err
+	}
+	return repo, nil
+}
+
+func convertToBareRepo(repo *git.Repository, repoPath string) error {
+	// Convert to a bare repo.
+	config, err := repo.Config()
+	if err != nil {
+		return err
+	}
+	config.Core.IsBare = true
+	repo.SetConfig(config)
+
+	tempDir, err := ioutil.TempDir("", "fgs")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	// Move "<REPO>/.git" directory to a temporary directory.
+	err = os.Rename(filepath.Join(repoPath, ".git"), filepath.Join(tempDir, ".git"))
+	if err != nil {
+		return err
+	}
+
+	// Delete <REPO> folder. This takes care of deleting all worktree files.
+	err = os.RemoveAll(repoPath)
+	if err != nil {
+		return err
+	}
+
+	// Move the .git folder to the <REPO> folder path.
+	err = os.Rename(filepath.Join(tempDir, ".git"), repoPath)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runSetupScript(repoPath, script string) error {
+	// Catch errors in the script.
+	script = "set -eu;" + script
+
+	logrus.Infof("setup script looks like: %v", script)
+
+	cmd := exec.Command("sh", "-c", script)
+	cmd.Dir = repoPath
+
+	// By default, make it so that the git commands contained in the script
+	// result in reproducible commits. This can be overridden by the script
+	// itself if it chooses to (re-)export the same environment variables.
+	cmd.Env = []string{
+		"GIT_AUTHOR_NAME=abc",
+		"GIT_AUTHOR_EMAIL=d@e.f",
+		"GIT_AUTHOR_DATE='Thu May 19 12:34:56 2022 +0000'",
+		"GIT_COMMITTER_NAME=abc",
+		"GIT_COMMITTER_EMAIL=d@e.f",
+		"GIT_COMMITTER_DATE='Thu May 19 12:34:56 2022 +0000'"}
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setGitConfigOptions(r *git.Repository) error {
+	config, err := r.Config()
+	if err != nil {
+		return err
+	}
+
+	// Ensure that the given Git repo allows anonymous push access. This is
+	// required for unauthenticated clients to push to the repo over HTTP.
+	config.Raw.SetOption("http", "", "receivepack", "true")
+
+	// Advertise all objects. This allows clients to fetch by raw commit SHAs,
+	// avoiding the dreaded
+	//
+	// 		Server does not allow request for unadvertised object <SHA>
+	//
+	// error.
+	config.Raw.SetOption("uploadpack", "", "allowAnySHA1InWant", "true")
+
+	r.SetConfig(config)
+
+	return nil
+}
+
+func defaultSignature() *object.Signature {
+	when, err := time.Parse(object.DateFormat, "Thu May 19 12:34:56 2022 +0000")
+	if err != nil {
+		logrus.Fatal(err)
```

Comment:
Forgot to delete this (it was used with the go-git library when creating Git commits with that library). Because we let clients create commits using `git` directly, we don't need this any more. Deleted.